File: neet_code_all_cpp/bin_search/search_insert_position.cpp
#include <iostream>
#include <vector>

// AC
class Solution {
    public:
        int searchInsert(std::vector<int>& nums, int target) {
            auto [l, r] = std::pair(0, (int)nums.size() - 1);
            auto mid = 0;

            while(l <= r) {
                mid = l + (r - l) / 2;
                if(target == nums[mid]) {
                    return mid;
                }

                if(nums[mid] < target) {
                    l = mid + 1;
                } else {
                    r = mid - 1;
                }
            }

            if(nums[mid] < target) {
                ++mid;
            }

            return mid;
        }
};

// AC
// 模範解答
class SolutionAns {
    public:
        int searchInsert(std::vector<int>& nums, int target) {
            int left = 0;
            int right = nums.size() - 1;

            while(left <= right) {
                int mid = left + (right - left) / 2;

                if(nums[mid] < target) {
                    left = mid + 1;
                }else {
                    right = mid - 1;
                }
            }

            return left;
        }
};

int main(void) {
    using Case = std::pair<std::vector<int>, int>;

    Case case_1 = {{1, 3, 5, 6}, 5};
    // => 2
    Case case_2 = {{1, 3, 5, 6}, 2};
    // => 1
    Case case_3 = {{1, 3, 5, 6}, 7};
    // => 4
    Case case_4 = {{1, 3, 5, 6}, 0};
    // => 0
    Case case_5 = {{3, 4, 5, 6, 7, 8}, 6};
    // => 3

    Solution s_1;
    std::cout << s_1.searchInsert(case_1.first, case_1.second) << std::endl;
    std::cout << s_1.searchInsert(case_2.first, case_2.second) << std::endl;
    std::cout << s_1.searchInsert(case_3.first, case_3.second) << std::endl;
    std::cout << s_1.searchInsert(case_4.first, case_4.second) << std::endl;
    std::cout << s_1.searchInsert(case_5.first, case_5.second) << std::endl;

    SolutionAns s_ans;
    std::cout << s_ans.searchInsert(case_1.first, case_1.second) << std::endl;
    std::cout << s_ans.searchInsert(case_2.first, case_2.second) << std::endl;
    std::cout << s_ans.searchInsert(case_3.first, case_3.second) << std::endl;
    std::cout << s_ans.searchInsert(case_4.first, case_4.second) << std::endl;
    std::cout << s_ans.searchInsert(case_5.first, case_5.second) << std::endl;
}

File: neet_code_all/src/bin_search/search_insert_position.rs
// AC
struct Solution {}
impl Solution {
    pub fn search_insert(nums: Vec<i32>, target: i32) -> i32 {
        let (mut l, mut r) = (0, nums.len() as i32 - 1);

        let mut mid = 0;
        while l <= r {
            mid = l + (r - l) / 2;
            if target == nums[mid as usize] {
                return mid;
            }

            if target < nums[mid as usize] {
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }

        //println!("not found");
        if nums[mid as usize] < target {
            return mid + 1;
        }

        mid
    }
}

// 模範解答
struct SolutionAns {}
impl SolutionAns {
    pub fn search_insert(nums: Vec<i32>, target: i32) -> i32 {
        match nums.binary_search(&target) {
            Ok(i) => i as i32,
            Err(i) => i as i32,
        }
    }
}

// AC
// C++の模範解答より
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    pub fn search_insert(nums: Vec<i32>, target: i32) -> i32 {
        let (mut left, mut right) = (0, nums.len() as i32 - 1);

        while left <= right {
            let mid = left + (right - left) / 2;

            if nums[mid as usize] < target {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        left
    }
}

fn main() {
    let case_1 = (vec![1, 3, 5, 6], 5);
    // => 2
    let case_2 = (vec![1, 3, 5, 6], 2);
    // => 1
    let case_3 = (vec![1, 3, 5, 6], 7);
    // => 4
    let case_4 = (vec![1, 3, 5, 6], 0);
    // => 0

    println!(
        "case_1: {}",
        Solution::search_insert(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        Solution::search_insert(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        Solution::search_insert(case_3.0.clone(), case_3.1)
    );
    println!(
        "case_4: {}",
        Solution::search_insert(case_4.0.clone(), case_4.1)
    );

    println!(
        "case_1: {}",
        SolutionAns::search_insert(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        SolutionAns::search_insert(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        SolutionAns::search_insert(case_3.0.clone(), case_3.1)
    );
    println!(
        "case_4: {}",
        SolutionAns::search_insert(case_4.0.clone(), case_4.1)
    );

    println!(
        "case_1: {}",
        SolutionAnsCpp::search_insert(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        SolutionAnsCpp::search_insert(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        SolutionAnsCpp::search_insert(case_3.0.clone(), case_3.1)
    );
    println!(
        "case_4: {}",
        SolutionAnsCpp::search_insert(case_4.0.clone(), case_4.1)
    );
}
